File: src/test/java/com/fwdekker/randomness/ui/JEditableListTest.java
package com.fwdekker.randomness.ui;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.NoSuchElementException;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;


/**
 * Unit tests for {@link JEditableList}.
 */
final class JEditableListTest {
    private JEditableList<String> list;


    @BeforeEach
    void beforeEach() {
        list = new JEditableList<>();
    }


    @Test
    void testGetEntriesEmpty() {
        assertThat(list.getEntries())
            .isEmpty();
    }

    @Test
    void testAddEntry() {
        list.addEntry("sxTODMrLvE");

        assertThat(list.getEntries())
            .containsExactly("sxTODMrLvE");
    }

    @Test
    void testAddEntryDuplicate() {
        list.addEntry("6bqmI9JEkv");
        list.addEntry("6bqmI9JEkv");

        assertThat(list.getEntries())
            .containsExactly("6bqmI9JEkv");
    }

    @Test
    void testSetEntriesEmptyToEmpty() {
        list.setEntries(Collections.emptyList());

        assertThat(list.getEntries())
            .isEmpty();
    }

    @Test
    void testSetEntriesEmptyToNonEmpty() {
        list.setEntries(Arrays.asList("[qOBxjKQ6P", "DGw{4ag(99"));

        assertThat(list.getEntries())
            .containsExactly("[qOBxjKQ6P", "DGw{4ag(99");
    }

    @Test
    void testSetEntriesNonemptyToEmpty() {
        list.addEntry("u1}9]G<<CN");
        list.addEntry("SV9MUxo5yM");
        list.setEntries(Collections.emptyList());

        assertThat(list.getEntries())
            .isEmpty();
    }

    @Test
    void testSetEntriesNonemptyToNonempty() {
        list.addEntry("d>OO>rp2zJ");
        list.addEntry("Po(tiIXnxQ");
        list.addEntry("oq5BQYk0<7");
        list.setEntries(Arrays.asList("qKZK]8Y[vs", "cn3{[6y>de", "skE}h6H9MJ"));

        assertThat(list.getEntries())
            .containsExactly("qKZK]8Y[vs", "cn3{[6y>de", "skE}h6H9MJ");
    }

    @Test
    void testRemoveEntry() {
        list.addEntry("><t1wfkCLz");
        list.addEntry("X0hxkJGK)7");
        list.addEntry("TI5i9bUyLc");
        list.removeEntry("X0hxkJGK)7");

        assertThat(list.getEntries())
            .containsExactly("><t1wfkCLz", "TI5i9bUyLc");
    }

    @Test
    void testRemoveEntryDuplicate() {
        list.addEntry("UNIODenA]8");
        list.addEntry(">q(isDDjGx");
        list.addEntry("tyFoveRt5n");
        list.addEntry("tyFoveRt5n");
        list.removeEntry("tyFoveRt5n");

        assertThat(list.getEntries())
            .containsExactly("UNIODenA]8", ">q(isDDjGx");
    }

    @Test
    void testRemoveEntryNonExisting() {
        list.addEntry("qiGIG5Slmn");
        list.addEntry("uxlt{9}FeZ");
        list.addEntry("a4BEQoJpLJ");

        assertThatThrownBy(() -> list.removeEntry("6uy19G<}JS"))
            .isInstanceOf(NoSuchElementException.class)
            .hasMessage("No row with entry `6uy19G<}JS` found.")
            .hasNoCause();
    }

    @Test
    void testClearEmpty() {
        list.clear();

        assertThat(list.getEntries())
            .isEmpty();
    }

    @Test
    void testClearNonempty() {
        list.addEntry("4lbw4K>}UH");
        list.addEntry("CMVe{rkBl[");
        list.clear();

        assertThat(list.getEntries())
            .isEmpty();
    }

    @Test
    void testGetEntryCountEmpty() {
        assertThat(list.getEntryCount())
            .isEqualTo(0);
    }

    @Test
    void testGetEntryCount() {
        list.addEntry("1JfHiG[CiD");
        list.addEntry("gXsoY21wzb");

        assertThat(list.getEntryCount())
            .isEqualTo(2);
    }

    @Test
    void testGetEntryCountDuplicate() {
        list.addEntry("jDwxzlHh]f");
        list.addEntry("A>)6q<f5kT");
        list.addEntry("A>)6q<f5kT");

        assertThat(list.getEntryCount())
            .isEqualTo(2);
    }


    @Test
    void testActiveEntriesEmpty() {
        assertThat(list.getActiveEntries())
            .isEmpty();
    }

    @Test
    void testActiveEntriesNone() {
        list.setEntries(Arrays.asList("DnqtROf4fd", "0<16CDsby2", "h0hu4XePhv"));

        assertThat(list.getActiveEntries())
            .isEmpty();
    }

    @Test
    void testActiveEntriesSome() {
        final List<String> entries = Arrays.asList("9}juZWluy}", "UGuD08qUXr", "eQA[]AdpYR");
        list.setEntries(entries);
        list.setActiveEntries(Arrays.asList("UGuD08qUXr"));

        assertThat(list.getActiveEntries())
            .containsExactly("UGuD08qUXr");
    }

    @Test
    void testActiveEntriesAll() {
        final List<String> entries = Arrays.asList("o28ix>b}x(", "6Zzl>yi5LB", "XyEVdjv1VM");
        list.setEntries(entries);
        list.setActiveEntries(entries);

        assertThat(list.getActiveEntries())
            .containsExactlyElementsOf(entries);
    }

    @Test
    @Disabled // TODO Re-enable this test. Should it throw an exception or not?
    void testSetActiveEntriesNonExistent() {
        final List<String> entries = Arrays.asList("JXIPoWsGR{", ">Jq7ILgv9]");
        list.setEntries(entries);
        list.setActiveEntries(Arrays.asList("JXIPoWsGR{", "j>NBYo}DnW", ">Jq7ILgv9]"));

        assertThat(list.getActiveEntries())
            .containsExactlyElementsOf(entries);
    }

    @Test
    void testIsActive() {
        final List<String> entries = Arrays.asList("I85kO5f8}6", "qcSv3u((zE", "{jjD)iFxEr");
        list.setEntries(entries);
        list.setEntryActivity("I85kO5f8}6", true);

        assertThat(list.isActive("I85kO5f8}6")).isTrue();
        assertThat(list.isActive("qcSv3u((zE")).isFalse();
        assertThat(list.isActive("{jjD)iFxEr")).isFalse();
    }

    @Test
    void testIsActiveNonExistent() {
        assertThatThrownBy(() -> list.isActive("Vr{1zIC9iH"))
            .isInstanceOf(NoSuchElementException.class)
            .hasMessage("No row with entry `Vr{1zIC9iH` found.")
            .hasNoCause();
    }

    @Test
    void testActivityListenerOnUpdate() {
        final boolean[] fired = {false};
        list.addEntry("DsX[DtA>6{");

        list.addEntryActivityChangeListener(event -> {
            fired[0] = true;
            return Unit.INSTANCE;
        });
        list.setEntryActivity("DsX[DtA>6{", true);

        assertThat(fired[0]).isTrue();
    }

    @Test
    void testActivityListenerNoChangeOnInsert() {
        final boolean[] fired = {false};

        list.addEntryActivityChangeListener(event -> {
            fired[0] = true;
            return Unit.INSTANCE;
        });
        list.addEntry("DsX[DtA>6{");

        assertThat(fired[0]).isFalse();
    }

    @Test
    void testActivityListenerRemoveNoFire() {
        final boolean[] fired = {false};
        list.addEntry("bvDAPSZFG3");

        final Function1<Integer, Unit> changeListener = event -> {
            fired[0] = true;
            return Unit.INSTANCE;
        };
        list.addEntryActivityChangeListener(changeListener);
        list.removeEntryActivityChangeListener(changeListener);
        list.setEntryActivity("bvDAPSZFG3", true);

        assertThat(fired[0]).isFalse();
    }


    @Test
    void testGetHighlightedEntryNone() {
        assertThat(list.getHighlightedEntry()).isNull();
    }

    @Test
    void testGetHighlightedEntrySingle() {
        list.setEntries(Arrays.asList("Bb]CEbJlAD", "8QNk5l<]ln", "U5Hbo0whnn"));
        list.addRowSelectionInterval(0, 0);

        assertThat(list.getHighlightedEntry())
            .isEqualTo("Bb]CEbJlAD");
    }

    @Test
    void testGetHighlightedEntryMultiple() {
        list.setEntries(Arrays.asList("k<Goz2<IG9", "E4nRBR>wKG", "sCxbg}sfy("));
        list.addRowSelectionInterval(0, 0);
        list.addRowSelectionInterval(2, 2);

        assertThat(list.getHighlightedEntry())
            .isEqualTo("sCxbg}sfy(");
    }


    @Test
    void testGetColumnClasses() {
        assertThat(list.getColumnClass(0)).isEqualTo(Boolean.class);
        assertThat(list.getColumnClass(1)).isEqualTo(String.class);

        assertThatThrownBy(() -> list.getColumnClass(-1))
            .isInstanceOf(IllegalArgumentException.class)
            .hasMessage("JEditableList has only two columns.")
            .hasNoCause();
        assertThatThrownBy(() -> list.getColumnClass(2))
            .isInstanceOf(IllegalArgumentException.class)
            .hasMessage("JEditableList has only two columns.")
            .hasNoCause();
    }

    @Test
    void testIsCellEditable() {
        assertThat(list.isCellEditable(10, 5)).isFalse();
        assertThat(list.isCellEditable(7, 10)).isFalse();
        assertThat(list.isCellEditable(8, 2)).isFalse();
    }
}

File: src/test/kotlin/com/fwdekker/randomness/ui/JEditableListTest.kt
package com.fwdekker.randomness.ui

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.assertj.swing.edt.GuiActionRunner
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.describe
import org.jetbrains.spek.api.dsl.it
import java.util.NoSuchElementException


/**
 * Unit tests for [JCheckBoxList].
 */
object JEditableListTest : Spek({
    lateinit var list: JCheckBoxList<String>


    beforeEachTest {
        list = GuiActionRunner.execute<JCheckBoxList<String>> { JCheckBoxList() }
    }


    describe("getEntries") {
        it("returns nothing if no entries are added") {
            assertThat(list.entries).isEmpty()
        }
    }

    describe("addEntry") {
        it("adds a given entry") {
            GuiActionRunner.execute { list.addEntry("manscape") }

            assertThat(list.entries).containsExactly("manscape")
        }

        it("adds two given entries") {
            GuiActionRunner.execute {
                list.addEntry("osteomas")
                list.addEntry("imido")
            }

            assertThat(list.entries).containsExactly("osteomas", "imido")
        }

        it("does not add a given entry twice") {
            GuiActionRunner.execute {
                list.addEntry("bloomed")
                list.addEntry("bloomed")
            }

            assertThat(list.entries).containsExactly("bloomed")
        }
    }

    describe("setEntries") {
        it("empties an empty list") {
            GuiActionRunner.execute { list.setEntries(emptyList()) }

            assertThat(list.entries).isEmpty()
        }

        it("adds two entries to an empty list") {
            GuiActionRunner.execute { list.setEntries(listOf("gladding", "flecky")) }

            assertThat(list.entries).containsExactly("gladding", "flecky")
        }

        it("does not add duplicate entries to a list") {
            GuiActionRunner.execute { list.setEntries(listOf("orrery", "orrery")) }

            assertThat(list.entries).containsExactly("orrery")
        }

        it("empties a non-empty list") {
            GuiActionRunner.execute {
                list.addEntry("prefeast")
                list.addEntry("strip")
            }

            GuiActionRunner.execute { list.setEntries(emptyList()) }

            assertThat(list.entries).isEmpty()
        }

        it("replaces the entries in a non-empty list") {
            GuiActionRunner.execute {
                list.addEntry("pirates")
                list.addEntry("underbit")
                list.addEntry("flexured")
            }

            GuiActionRunner.execute { list.setEntries(listOf("steevely", "qual", "wheaties")) }

            assertThat(list.entries).containsExactly("steevely", "qual", "wheaties")
        }
    }

    describe("removeEntry") {
        it("removes the given entry") {
            GuiActionRunner.execute {
                list.addEntry("giftwrap")
                list.addEntry("backbite")
                list.addEntry("landwhin")
            }

            GuiActionRunner.execute { list.removeEntry("backbite") }

            assertThat(list.entries).containsExactly("giftwrap", "landwhin")
        }

        it("throws an exception if the element to be removed cannot be found") {
            GuiActionRunner.execute {
                list.addEntry("tracheid")
                list.addEntry("cocomat")
                list.addEntry("videotex")
            }

            assertThatThrownBy { list.removeEntry("unshowed") }
                .isInstanceOf(NoSuchElementException::class.java)
                .hasMessage("No row with entry `unshowed` found.")
                .hasNoCause()
        }
    }

    describe("clear") {
        it("clears an empty list") {
            GuiActionRunner.execute { list.clear() }

            assertThat(list.entries).isEmpty()
        }

        it("clears a non-empty list") {
            GuiActionRunner.execute {
                list.addEntry("lathered")
                list.addEntry("aquench")
            }

            GuiActionRunner.execute { list.clear() }

            assertThat(list.entries).isEmpty()
        }
    }

    describe("entryCount") {
        it("counts 0 elements for an empty list") {
            assertThat(list.entryCount).isEqualTo(0)
        }

        it("counts 2 elements for a list with 2 elements") {
            GuiActionRunner.execute {
                list.addEntry("musmon")
                list.addEntry("topepo")
            }

            assertThat(list.entryCount).isEqualTo(2)
        }
    }

    describe("activeEntries") {
        it("has no active entries in an empty list") {
            assertThat(list.activeEntries).isEmpty()
        }

        it("has no active entries by default") {
            GuiActionRunner.execute { list.setEntries(listOf("otxi", "chamorro", "scarab")) }

            assertThat(list.activeEntries).isEmpty()
        }

        it("can enable one entry") {
            GuiActionRunner.execute { list.setEntries(listOf("disbury", "curet", "sunhat")) }

            GuiActionRunner.execute { list.setActiveEntries(listOf("curet")) }

            assertThat(list.activeEntries).containsExactly("curet")
        }

        it("can enable all entries") {
            val entries = listOf("big", "coumaric", "hooray")
            GuiActionRunner.execute { list.setEntries(entries) }

            GuiActionRunner.execute { list.setActiveEntries(entries) }

            assertThat(list.activeEntries).containsExactlyElementsOf(entries)
        }

        it("ignores when a non-existing element is enabled") {
            val entries = listOf("forte", "blarneys")
            GuiActionRunner.execute { list.setEntries(entries) }

            GuiActionRunner.execute { list.setActiveEntries(listOf("forte", "pittings")) }

            assertThat(list.activeEntries).containsExactly("forte")
        }

        it("unchecks previously checked entries") {
            val entries = listOf("rot", "possible", "harbor")
            GuiActionRunner.execute { list.setEntries(entries) }
            GuiActionRunner.execute { list.setActiveEntries(entries) }

            GuiActionRunner.execute { list.setActiveEntries(listOf("rot", "harbor")) }

            assertThat(list.activeEntries).containsExactly("rot", "harbor")
        }
    }

    describe("isActive") {
        it("returns true iff an element is active") {
            GuiActionRunner.execute { list.setEntries(listOf("overtoil", "gouramis", "abruptly")) }

            GuiActionRunner.execute { list.setEntryActivity("overtoil", true) }

            assertThat(list.isActive("overtoil")).isTrue()
            assertThat(list.isActive("gouramis")).isFalse()
            assertThat(list.isActive("abruptly")).isFalse()
        }

        it("throws an exception if an element is not in the list") {
            assertThatThrownBy { list.isActive("nanism") }
                .isInstanceOf(NoSuchElementException::class.java)
                .hasMessage("No row with entry `nanism` found.")
                .hasNoCause()
        }
    }

    describe("activityListener") {
        it("fires when an entry's activity is changed") {
            var fired = false
            GuiActionRunner.execute { list.addEntry("scarpa") }
            list.addEntryActivityChangeListener { fired = true }

            GuiActionRunner.execute { list.setEntryActivity("scarpa", true) }

            assertThat(fired).isTrue()
        }

        it("does not fire when an entry is inserted") {
            var fired = false
            list.addEntryActivityChangeListener { fired = true }

            GuiActionRunner.execute { list.addEntry("techiest") }

            assertThat(fired).isFalse()
        }

        it("stops firing after a listener is removed") {
            var fired = false
            val listener = { _: Int -> fired = true }
            GuiActionRunner.execute { list.addEntry("optimum") }
            list.addEntryActivityChangeListener(listener)

            list.removeEntryActivityChangeListener(listener)
            GuiActionRunner.execute { list.setEntryActivity("optimum", true) }

            assertThat(fired).isFalse()
        }
    }

    describe("getHighlightedEntry") {
        it("returns null when no entry is highlighted") {
            assertThat(list.highlightedEntry).isNull()
        }

        it("returns the single highlighted entry") {
            GuiActionRunner.execute {
                list.setEntries(listOf("bahay", "woodyard", "hussies"))
                list.addRowSelectionInterval(0, 0)
            }

            assertThat(list.highlightedEntry).isEqualTo("bahay")
        }

        it("returns the first highlighted entry") {
            GuiActionRunner.execute {
                list.setEntries(listOf("shutoffs", "dentine", "scutel"))
                list.addRowSelectionInterval(0, 0)
                list.addRowSelectionInterval(2, 2)
            }

            assertThat(list.highlightedEntry).isEqualTo("scutel")
        }
    }

    describe("getColumnClass") {
        it("returns the right types for the columns") {
            // Java classes MUST be used
            assertThat(list.getColumnClass(0)).isEqualTo(java.lang.Boolean::class.java)
            assertThat(list.getColumnClass(1)).isEqualTo(java.lang.String::class.java)
        }

        it("throws an exception if a negative column index is requested") {
            assertThatThrownBy { list.getColumnClass(-1) }
                .isInstanceOf(IllegalArgumentException::class.java)
                .hasMessage("JEditableList has only two columns.")
                .hasNoCause()
        }

        it("throws an exception if column index that is too high is requested") {
            assertThatThrownBy { list.getColumnClass(2) }
                .isInstanceOf(IllegalArgumentException::class.java)
                .hasMessage("JEditableList has only two columns.")
                .hasNoCause()
        }
    }

    describe("isCellEditable") {
        it("returns true iff the column is 0") {
            assertThat(list.isCellEditable(13, 0)).isTrue()
            assertThat(list.isCellEditable(67, 0)).isTrue()

            assertThat(list.isCellEditable(10, 5)).isFalse()
            assertThat(list.isCellEditable(7, 10)).isFalse()
            assertThat(list.isCellEditable(8, 2)).isFalse()
        }
    }
})
